### PR TITLE
[WIP]Subsystemized AStar

### PR DESCRIPTION
**Pull request by silicons**
Moves AStar to its own subsystem so it can be tracked and killed if necessary
Wip:
Converting all existing uses of AStar to work with new procs as they're async.

### CODE REVIEW

File: code/__HELPERS/AStar.dm [hunk 1]
```diff
@@ -60,94 +60,142 @@ Actual Adjacent procs :
...
+				P.nt = cur.nt + 1
+				open.ReSort(P)	//Reorder the changed element in the list
+
+/datum/AStar/proc/Finish()
+	finished = TRUE
+	ready = FALSE
+	processing = FALSE
+	pnodelist = null	//Clean up
+	if(path)			//Reverse path for start to finish.
 		for(var/i = 1; i <= path.len/2; i++)
```
Comment (Iamgoofball) [inline code/__HELPERS/AStar.dm]: fix this to work with our standards for a performance boost

File: code/controllers/subsystem/processing/pathing.dm [hunk 2]
```diff
@@ -0,0 +1,12 @@
+
+//Intended for pathfinding operations like AStar
+
+PROCESSING_SUBSYSTEM_DEF(pathing)
+	name = "Pathfinding"
+	wait = 1
+	priority = 80		//No one cares about laggy pathfinding!
```
Comment (Iamgoofball) [inline code/controllers/subsystem/processing/pathing.dm]: ahem, we do actually

File: code/__HELPERS/AStar.dm [hunk 3]
```diff
@@ -60,94 +60,142 @@ Actual Adjacent procs :
...
-			path.Swap(i,path.len-i+1)
-
-	return path
+			path.Swap(i, path.len-i+1)
+	sendPath()
+	return PROCESS_KILL
```
Comment (Iamgoofball) [inline code/__HELPERS/AStar.dm]: have you tested the shit out of all this

**Comment by Iamgoofball** [comment]
We are absolutely going to need detailed and EXACT profiling results along with timestamps before this is approved, provide ASAP

**Comment by Iamgoofball** [comment]
@MrStonedOne @Cyberboss @coiax @Giacom @MrPerson iirc you guys are experts on this meme

**Comment by kevinz000** [comment]
....
....
How the fuck do I profile this?

**Comment by kevinz000** [comment]
Also I need to make the subsystem use all available cpu to calculate instead of only calculating one step..

**Comment by kevinz000** [comment]
Actually would it be better if it's still the old loop but there was a var to terminate it?

**Comment by Giacom** [comment]
Async pathfinding is hard so good luck. Only comment is that now you're returning a big datum, which is probably going to be slow to allocate / deallocate. I'm guessing thats to make it work asynchronously so its a common tricky question of whether it is faster to do the work yourself or to do the work that gets it done later. I also hope everything that uses Astar doesn't block.

**Comment by optimumtact** [comment]
LMAO good luck

**Comment by optimumtact** [comment]
@Giacom we already do astar async now (sort of, it can sleep via CHECK_TICK and then return to generating the path)

this is kinda hacky in that we never bother to verify our path is still correct but yolo

**Pull request closed without merging**